src/stream/src/executor/dml.rs

```diff
@@ -120,12 +149,62 @@ impl DmlExecutor {
                                 _ => {}
                             }
                         }
+
+                        // Stop barrier. It could be issued by scaling or DDLs(e.g. add column).
+                        if barrier.is_stop(self.actor_id) {
+                            self.dml_manager
+                                .unregister_changes_sender(self.table_id, self.actor_id);
+                        }
                     }
                     yield msg;
                 }
-                Either::Right(chunk) => {
+                Either::Right(txn_msg) => {
                     // Batch data.
-                    yield Message::Chunk(chunk.chunk);
+                    match txn_msg {
+                        TxnMsg::Begin(txn_id) => {
+                            active_txn_map
+                                .try_insert(txn_id, TxnBuffer::default())
+                                .unwrap_or_else(|_| {
+                                    panic!("Transaction id collision txn_id = {}.", txn_id)
+                                });
+                        }
+                        TxnMsg::End(txn_id) => {
+                            let mut txn_buffer = active_txn_map.remove(&txn_id)
+                                .unwrap_or_else(|| panic!("Receive an unexpected transaction end message. Active transaction map doesn't contain this transaction txn_id = {}.", txn_id));
+                            for chunk in txn_buffer.vec.drain(..) {
+                                yield Message::Chunk(chunk);
+                            }
+                        }
+                        TxnMsg::Rollback(txn_id) => {
+                            let txn_buffer = active_txn_map.remove(&txn_id)
+                                .unwrap_or_else(|| panic!("Receive an unexpected transaction rollback message. Active transaction map doesn't contain this transaction txn_id = {}.", txn_id));
+                            if txn_buffer.overflow {
+                                tracing::warn!("txn_id={} large transaction tries to rollback, but part of its data has already been sent to the downstream.", txn_id);
+                            }
+                        }
+                        TxnMsg::Data(txn_id, chunk) => {
```

Comment:
Do we have some strategies to prevent a long-running transaction from happening, or if the `End` message is somehow dropped (though seems impossible normally unless the batch runtime crashes without releasing the resources)?

Reply:
Previously I tried to remove the `TxnBuffer` if it exists for a long time, but since our transaction right now is a single statement which is not an interactive transaction, I think we can treat a long-running transaction as a large transaction currently. We have `MAX_CHUNK_FOR_ATOMICITY` to limit the large transaction buffer size. As for whether the `End` message is somehow missing, I think it is impossible because all the communication between the DML statement and the dml executor is in the same node. At least it will send `Rollback` if failure occurs.

Reply:
> I think we can treat a long-running transaction as a large transaction currently

We should also expect users not to issue a complicated batch query as the input of the DML executors.